### PR TITLE
Provide mirror support after inlining

### LINKED ISSUE #11542: derivation: capturing mirror and nested compile-time summoning leads to ClassCastException at runtime
## Compiler version

3.0.0-RC1

## Minimized code

```Scala
object demo {

  trait Reader[A]

  given Reader[Int] with {}

  inline def summonReader[T <: Tuple]: List[Reader[_]] = inline compiletime.erasedValue[T] match {
    case _: EmptyTuple => Nil
    case _: (t *: ts) => compiletime.summonInline[Reader[t]] :: summonReader[ts]
  }

  class CombinedReader[A](
    m: deriving.Mirror.ProductOf[A],
    childReaders: List[Reader[_]]
  ) extends Reader[A]

  inline given [A <: Tuple](using m: deriving.Mirror.ProductOf[A]): Reader[A] = {
    new CombinedReader(m, summonReader[m.MirroredElemTypes])
  }

}

@main def run() = {
  // OK
  //summon[demo.Reader[(Int, Int, Int)]]

  // Exception in thread "main" java.lang.ClassCastException: class main$package$$anon$2 cannot be cast to class scala.deriving.Mirror$Product (main$package$$anon$2 and scala.deriving.Mirror$Product are in unnamed module of loader 'app')
  //      at main$package$.run(main.scala:25)
  //      at run.main(main.scala:23)
  summon[demo.Reader[(Int, (Int, Int))]]
}

```

## Output

The program compiles, but encounters a ClassCastException at runtime, as shown in the comment above.

## Expectation

There should be no exception.

## Notes

The class cast exception only happens if the mirror is captured *and* there is a nested summoning. If, for example, `CombinedReader` would not use take `m` or `childReader` arguments, the program works as expected.

### LINKED ISSUE #11961: ClassCastException when trying to use Mirror.Of[T] when T contains a Tuple
## Compiler version

3.0.0-RC2
sbt 1.4.9 (GraalVM Community Java 1.8.0_282)

## Minimized code

```Scala
import scala.deriving.*
import scala.compiletime.{erasedValue, summonInline}

case class Simple(a: String, b: Boolean) derives Printable
case class SimpleT(a: (String, Boolean)) derives Printable

@main def hello: Unit = {

  summon[Printable[Simple]].print // Prints STRING BOOLEAN as expected

  summon[Printable[SimpleT]].print // java.lang.ClassCastException: SimpleT$$anon$1 cannot be cast to scala.deriving.Mirror$Product

}

trait Printable[T]:
   def print: Unit

object Printable:

   given Printable[String] with
      def print: Unit = println("STRING")

   given Printable[Boolean] with
      def print: Unit = println("BOOLEAN")

   def printProduct[T](p: Mirror.ProductOf[T], elems: => List[Printable[_]]): Printable[T] =
      new Printable[T]:
         def print: Unit =
            elems.foreach(_.print)

   inline given derived[T](using m: Mirror.Of[T]): Printable[T] =
      val elemInstances = summonAllPrintable[m.MirroredElemTypes]
      inline m match
         case p: Mirror.ProductOf[T] => printProduct(p, elemInstances)

end Printable

inline def summonAllPrintable[T <: Tuple]: List[Printable[_]] =
   inline erasedValue[T] match
      case _: EmptyTuple => Nil
      case _: (t *: ts) => summonInline[Printable[t]] :: summonAllPrintable[ts]
```

## Output

```
> run
[info] running hello 
STRING
BOOLEAN
[error] (run-main-0) java.lang.ClassCastException: SimpleT$$anon$1 cannot be cast to scala.deriving.Mirror$Product
[error] java.lang.ClassCastException: SimpleT$$anon$1 cannot be cast to scala.deriving.Mirror$Product
[error] 	at SimpleT$.derived$Printable(Main.scala:5)
[error] 	at Main$package$.hello(Main.scala:11)
[error] 	at hello.main(Main.scala:7)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.lang.reflect.Method.invoke(Method.java:498)
[error] stack trace is suppressed; run last Compile / bgRun for the full output
[error] Nonzero exit code: 1
[error] (Compile / run) Nonzero exit code: 1
[error] Total time: 3 s, completed Mar 31, 2021 9:16:29 PM
```

## Expectation
I'd expect this program to produce compilation error about missing `Printable[(String, Boolean)]` or similar.

Trying to get familiar with [type-classes-supporting-automatic-deriving](https://dotty.epfl.ch/docs/reference/contextual/derivation.html#type-classes-supporting-automatic-deriving) I'm trying to implement typeclass which prints its own type.

Minimized project is available [here](https://github.com/VlachJosef/scala3-tuple-cast-exception).

### LINKED ISSUE #12052: Summoning a mirror for a Tuple produces a class-cast-exception during runtime
## Compiler version

Latest nightly (i.e. `dottyLatestNightlyBuild.get`)

## Minimized code

Create a simple mirror type object which summons a product mirror, add an extension method for it to be able to be applied to any type:
```scala
object MirrorType {
  class Container[T]

  inline def decode[T]: String =
    summonFrom {
      case ev: Mirror.ProductOf[T] =>
        s"Product-${new Container[ev.MirroredElemLabels]}" // This is the part that splices in the cast
      case m: Mirror.SumOf[T] =>
        "Sum"
    }

  inline def generic[T]: MirrorType[T] = 
    new MirrorType[T] {
      def mirrorType: String = decode[T]
    }

  extension[T](inline value: T)
    inline def mirrorType = summonFrom {
      case mt: MirrorType[T] => mt.mirrorType
      case _ => "mirror not found"
    }
}

```
Then import the context and run it:
```scala
object ContextUse {
  def main(args: Array[String]): Unit = {
    val ctx = new MyContext();
    import ctx._
    val tup = ("foo", 1)
    println( tup.mirrorType )
  }
}
```

## Output

A class cast exception will occur
```
[error] (run-main-2) java.lang.ClassCastException: io.getquill.mytest.ContextUse$$anon$2 cannot be cast to scala.deriving.Mirror$Product
[error] java.lang.ClassCastException: io.getquill.mytest.ContextUse$$anon$2 cannot be cast to scala.deriving.Mirror$Product
[error]         at io.getquill.mytest.ContextUse$$anon$1.mirrorType(Use.scala:10)
[error]         at io.getquill.mytest.ContextUse$.main(Use.scala:10)
[error]         at io.getquill.mytest.ContextUse.main(Use.scala)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]         at java.lang.reflect.Method.invoke(Method.java:498)
```

This is probably because I can see that the following code is being spliced:
```scala
{
  val ev: Product {
    type MirroredType >: Tuple2[String, Int] <: Tuple2[String, Int]
    type MirroredMonoType >: Tuple2[String, Int] <: Tuple2[String, Int]
    type MirroredElemTypes >: Nothing <: Tuple
  } & Product {
    type MirroredMonoType >: Tuple2[String, Int] <: Tuple2[String, Int]
    type MirroredType >: Tuple2[String, Int] <: Tuple2[String, Int]
    type MirroredLabel >: "Tuple2" <: "Tuple2"
  } {
    type MirroredElemTypes >: *:[String, *:[Int, EmptyTuple]] <: *:[String, *:[Int, EmptyTuple]]
    type MirroredElemLabels >: *:["_1", *:["_2", EmptyTuple]] <: *:["_1", *:["_2", EmptyTuple]]
  } = {
    final class $anon() {
      type MirroredMonoType = Tuple2[String, Int]
    }

    (new $anon(): Object)
  }.$asInstanceOf$[Product {      // WHAT??? You can't cast this to a Product
    type MirroredType >: Tuple2[String, Int] <: Tuple2[String, Int]
    type MirroredMonoType >: Tuple2[String, Int] <: Tuple2[String, Int]
    type MirroredElemTypes >: Nothing <: Tuple
  } & Product {
    type MirroredMonoType >: Tuple2[String, Int] <: Tuple2[String, Int]
    type MirroredType >: Tuple2[String, Int] <: Tuple2[String, Int]
    type MirroredLabel >: "Tuple2" <: "Tuple2"
  } {
    type MirroredElemTypes >: *:[String, *:[Int, EmptyTuple]] <: *:[String, *:[Int, EmptyTuple]]
    type MirroredElemLabels >: *:["_1", *:["_2", EmptyTuple]] <: *:["_1", *:["_2", EmptyTuple]]
  }]
  _root_.scala.StringContext.apply("Product-", "").s(new Container[MirroredElemLabels]())
}
```

## Expectation
Code should compile and run correctly.

## Repository
Code is available here:
https://github.com/deusaquilus/class_cast_issue

### PR DESCRIPTION
**Pull request by odersky**
Mirror support runs in PostTyper to add new members to mirrors generated
during Typer. But some anonymous mirrors are generated during inlining.
We need to add the missing methods for them as well.

Fixes #11542
Fixes #11961
Fixes #12052

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/PostInlining.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+/** A phase that adds mirror support for anonymous mirrors created at inlining. */
+class PostInlining extends MacroTransform, IdentityDenotTransformer:
+  thisPhase =>
+
+  override def phaseName: String = PostInlining.name
+  override def changesMembers = true
+
+  override def run(using Context): Unit =
+    if ctx.compilationUnit.needsMirrorSupport then super.run
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/PostInlining.scala:18]:
```suggestion
    if ctx.compilationUnit.needsMirrorSupport && ctx.compilationUnit.needsInlining then super.run
```

Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/PostInlining.scala:18]: We don't need that. needsMirrorSupport is only set of something was inlined.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PostInlining.scala:19]: We should document that in the definition of `needsMirrorSupport`. Future changes might easily break that assumption.

File: compiler/src/dotty/tools/dotc/CompilationUnit.scala [hunk 2]
```diff
@@ -43,6 +43,8 @@ class CompilationUnit protected (val source: SourceFile) {
    */
   var needsInlining: Boolean = false
 
+  var needsMirrorSupport: Boolean = false
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/CompilationUnit.scala:46]:
```suggestion
  /** Set to `true` if inliner added anonymous mirrors that need to be completed */
  var needsMirrorSupport: Boolean = false
```

**Review by nicolasstucki — APPROVED** [review]

**Pull request merged**